src/core/infer.ml

```diff
@@ -38,10 +41,102 @@ let unif : problem -> octxt -> term -> term -> unit =
+    true)
+  else (
+    Time.restore timestamp;
+    unif pb c a b;
+    !solve pb || (Time.restore timestamp; false))
+
 (** {1 Handling coercions} *)
 
+(** [is_value t] is true if term [t] does not contain [Coercions.coerce]. *)
+let is_value t =
+  let exception Found in
+  (* TODO: write a generic [iter] that operates on leaves *)
+  let rec is_value t =
+    match unfold t with
+    | TRef _-> assert false
+    | TEnv _-> assert false
+    | Wild -> assert false
+    | Patt _ -> assert false
+    | Plac _ -> assert false
+    | Type | Kind | Vari _ | Meta _ -> ()
+    | Symb s when Coercions.is_ghost s -> raise Found
+    | Symb _-> ()
+    | Appl (t, u) -> is_value t; is_value u
+    | Prod(a,b)
+    | Abst(a,b) ->
+        is_value a; is_value (snd (Bindlib.unbind b))
+    | LLet(a,t,u) ->
+        is_value a; is_value t; is_value (snd (Bindlib.unbind u))
+  in
+  try is_value t; true with Found -> false
+
+(** [unify_remains pb c t] traverses term [t] and for each subterm of
+    the form [#c a t b] it tries to [solve1 pb c a b] *)
+let unify_remains pb c t =
+  if Logger.log_enabled () then
+    log "Unifying remains of [%a]" Print.pp_term t;
+  let exception Failure in
+  let rec unre t =
+    match get_args t with
+    | TRef _, _-> assert false
+    | TEnv _, _ -> assert false
+    | Wild, _ -> assert false
+    | Patt _, _ -> assert false
+    | Plac _, _ -> assert false
+    | (Type | Kind | Vari _ | Meta _), _ -> ()
+    | Symb s, [a;_;b] when Coercions.is_ghost s ->
+        if not (solve1 pb c a b) then raise Failure
+    | Symb _, args -> List.iter unre args
+    | (Prod(a,b) | Abst(a,b)), args ->
+        unre a; unre (snd (Bindlib.unbind b)); List.iter unre args
+    | LLet(a,t,u), args ->
+        unre a; unre t; unre (snd (Bindlib.unbind u)); List.iter unre args
+    | Appl _, _ -> assert false
+  in
+  try unre t; true with Failure -> false
+
 (* Simply unify types, no coercions yet. *)
-let coerce pb c t a b = unif pb c a b; (t, false)
+let coerce pb c t a b =
+  if Eval.eq_modulo (classic c) a b then (t, false) else
+  (* FIXME: tests/OK/273.lp fails if this option isn't used (because when
+     no coercion is needed, unification constraints may be postponed) *)
+  if !(Coercions.coerce.sym_dtree) == Tree_type.empty_dtree then (
+    unif pb c a b;
+    (t, false))
+  else
+    if solve1 pb c a b then (t, false) else (
+      if Logger.log_enabled () then
+        log "Cast [%a: %a = %a]"
+          Print.pp_term t Print.pp_term a Print.pp_term b;
+      let reduced = ref (add_args (mk_Symb Coercions.coerce) [a; t; b]) in
+      let exception Over in
+      let exception Failure in
+      try
+        while true do
+          reduced := Eval.snf (classic c) !reduced;
+          if is_value !reduced then raise Over;
+          let size = MetaSet.cardinal !pb.metas in
+          ignore (unify_remains pb c !reduced);
+          if not ((MetaSet.cardinal !pb.metas) < size) then raise Failure
```

Comment:
Some explanation is important here: what do you do and why?

Reply:
Right it's a bit cryptic: it may happen that snf leaves terms of the form `#c ?0 t a`. In that case, of course `?0` is not convertible to `a`, so this cannot be simplified, even though we have the rule `#c $a $t $a --> $t`. Therefore we try to unify `$0` with `a`. If it succeeds, then `?0` is instantiated to `a` and next call to snf will reduce to `t`. The different steps of the loop are thus:
1. try to reduce,
2. try to unify what remains
3. if unification has instantiated some metavariables, we may be able to reduce further the coercions so repeat. If not, then probably no coercion can be found.